### PR TITLE
fix: add type declarations

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

Adds type declarations for the package.

## Context:

- Ref #112

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
@rarkins Could you please approve the workflows.

**Comment by RahulGautamSingh** [comment]
@rarkins Could you approve the workflows please..

**Comment by RahulGautamSingh** [comment]
@rarkins Could you approve the workflows again please..

**Comment by rarkins** [comment]
@viceice could you review?

File: lib/version.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+export declare const VERSION_PATTERN: string;
+export declare function valid(version: string): string | null;
+export declare function clean(version: string): string | null;
+export declare function stringify(version: string): string | null;
+
+interface parsed {
```
Comment (viceice) [inline lib/version.d.ts:6]: Do not duplicate, use a shared types file and import from it.

Comment (RahulGautamSingh) [reply lib/version.d.ts]: Done🟩
Added types for functions with similar parameters as well, needs better naming though.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import {
```
Comment (viceice) [inline index.d.ts:1]:
```suggestion
import type {
```

File: lib/operator.d.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { OperatorFunctionType } from "./shared";
```
Comment (viceice) [inline lib/operator.d.ts:1]:
```suggestion
import type { OperatorFunctionType } from "./shared";
```

File: lib/semantic.d.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { SemanticFunctionType } from "./shared";
```
Comment (viceice) [inline lib/semantic.d.ts:1]:
```suggestion
import type { SemanticFunctionType } from "./shared";
```

File: lib/shared.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { SemVer } from "semver";
```
Comment (viceice) [inline lib/shared.ts:1]:
```suggestion
import type { SemVer } from "semver";
```

File: lib/specifier.d.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { parsed, Range } from "./shared";
```
Comment (viceice) [inline lib/specifier.d.ts:1]:
```suggestion
import type { parsed, Range } from "./shared";
```

File: lib/version.d.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import type { VersionFunctionType, parsed } from "./shared";
```
Comment (viceice) [inline lib/version.d.ts:1]:
```suggestion
import type { VersionFunctionType, parsed } from "./shared";

```

**Comment by RahulGautamSingh** [comment]
Ready for review and merge.🟩

**Review by viceice — COMMENTED** [review]
Why are some functions have a declare? They should have all a declare or none.

**Comment by RahulGautamSingh** [comment]
> Why are some functions have a declare? They should have all a declare or none.

Resolved

**Comment by RahulGautamSingh** [comment]
Ready for review.
_For some reason the re-view btn is not working for this PR_

**Comment by RahulGautamSingh** [comment]
> Ready for review. _For some reason the request review btn is not working for this PR_

File: lib/specifier.d.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import type { parsed, Range } from "./shared";
+
+export declare const RANGE_PATTERN: string;
```
Comment (viceice) [inline lib/specifier.d.ts:3]:
```suggestion
export const RANGE_PATTERN: string;
```

File: lib/shared.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import type { SemVer } from "semver";
```
Comment (viceice) [inline lib/shared.ts:1]: Rename file extension to `.d.ts`

**Comment by viceice** [comment]
@rarkins can you check my permission on this repo?

**Comment by RahulGautamSingh** [comment]
Ready.

**Comment by viceice** [comment]
You need to add the `index.d.ts` to `files` in `package.json`

**Comment by RahulGautamSingh** [comment]
> You need to add the `index.d.ts` to `files` in `package.json`

done.
I thought of the same, like adding `index.d.ts` to `package.json` but when I tested locally without adding it worked fine.
Maybe in renovate we have configured it to work that way too. I forgot to add it back after the test.

**Comment by RahulGautamSingh** [comment]
> > You need to add the `index.d.ts` to `files` in `package.json`
> 
> done. I added `index.d.ts` to `package.json` initially but when I tested locally without adding it worked fine. Maybe in renovate we have configured it to work that way too. I forgot to add it back after the test.

Found the reason, it is not our config but a default behaviour, when `inedx.d.ts` files are used.
```Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the types property, though it is **advisable** to do so.```
[link](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html)

File: package.json [hunk 10]
```diff
@@ -51,7 +51,8 @@
   },
   "files": [
     "index.js",
-    "lib"
+    "lib",
+    "index.d.ts"
```
Comment (viceice) [inline package.json:55]: If this is missing, the file won't included in published package. 😉

**Review by viceice — APPROVED** [review]

File: package.json (hunk 10 above)
Comment (RahulGautamSingh) [reply package.json:55]: I read somewhere that all files are included by default now, and only the files under `.npmignore` are left out.

Nonetheless this is a good code-practice and an advisable thing to do as per the docs.✨

Comment (viceice) [reply package.json:55]: Yeah, by default all files are included. But if you define the `files` explicit, only those will be included in published package.

Comment (RahulGautamSingh) [reply package.json:55]: Ahh yes. I misunderstood and thought that even if we explicitly have a files property all others will be included.
Thanks , more clarity now.🙏

**Pull request merged**